api.go

```diff
@@ -71,26 +71,26 @@ func (conn *Connection) rqliteApiCall(apiOp apiOperation, method string, request
 			continue
 		}
 
-		// Check response code before reading body
-		if response.StatusCode != http.StatusOK {
-			trace("%s: got code %s", conn.ID, response.Status)
-			failureLog = append(failureLog, fmt.Sprintf("%s failed, got: %s", url, response.Status))
-			response.Body.Close()
-			continue
-		}
-
-		// Read response body now that we've got a successful answer
-		trace("%s: client.Do() OK", conn.ID)
+		// Read response body even if not a successful answer to return a descriptive error message
 		responseBody, err := ioutil.ReadAll(response.Body)
 		if err != nil {
 			trace("%s: got error '%s' doing ioutil.ReadAll", conn.ID, err.Error())
 			failureLog = append(failureLog, fmt.Sprintf("%s failed due to %s", url, err.Error()))
 			response.Body.Close()
 			continue
 		}
-		response.Body.Close()
 		trace("%s: ioutil.ReadAll() OK", conn.ID)
 
+		// Check that we've got a successful answer
+		if response.StatusCode != http.StatusOK {
+			trace("%s: got code %s", conn.ID, response.Status)
+			failureLog = append(failureLog, fmt.Sprintf("%s failed, got: %s, message: %s", url, response.Status, string(responseBody)))
+			response.Body.Close()
+			continue
+		}
+		response.Body.Close()
```

Comment:
Should we make this `defer response.Body.Close()` after `responseBody, err := ioutil.ReadAll(response.Body)` and remove the explicit calls to `response.Body.Close()` elsewhere?

Reply:
It wouldn't work the same since this is happening inside a for loop and the defer wouldn't run until the function returns.

Reply:
Ah yes.